src/wallet/wallet.cpp

```diff
@@ -1703,6 +1703,77 @@ int64_t CWallet::RescanFromTime(int64_t startTime, const WalletRescanReserver& r
     return startTime;
 }
 
+// Manages the log and UI progress display for CWallet::ScanForWalletTransactions
+class RescanVerificationProgress
+{
+private:
+    const CWallet& m_wallet;
+    int64_t m_time;
+    const CBlockIndex* m_current_block;
+    const CBlockIndex* m_stop_block;
+    double m_start;
+    double m_end;
+    double m_current;
+
+    double GuessProgress(const CBlockIndex* block)
+    {
+        AssertLockHeld(cs_main);
+        return GuessVerificationProgress(Params().TxData(), block);
+    }
+
+public:
+    RescanVerificationProgress(const CWallet& wallet, const CBlockIndex* const stop_block)
+        : m_wallet(wallet), m_time(GetTime()), m_stop_block(stop_block) {}
+
+    void Start(const CBlockIndex* const start_block)
```

Comment:
simpler to RAII this instead of a separate Constructor/Start/AnnounceStart methods?

Reply:
The main reason I kept this separate is that `GuessVerificationProgress` needs `cs_main` held, and I wanted to minimize the amount of code in the lock-block.